### DISCUSSION TITLE
Incorporate global descriptors into MACE

### DISCUSSION THREAD

**Discussion by pfebrer** [ACEsuit/mace#446] [Q&A] [answered]
Hi!

I have a global descriptor (a vector, the external electric field) that is very important for my predictions. What is the best way to incorporate this into MACE?

I have thought of incorporating the vector to all nodes in `node_attrs`, which would require a slight modification to accept `node_attrs_irreps` as an argument to initialize the model. I don't know if it's the best approach.

Is there some "established" way to do this?

Thanks!

**Comment by ilyes319** [comment, marked as answer]
Yes this is a good approach, just generalize to accept more than scalar `node_attrs`, it should be a small modification.

**Reply by pfebrer** [reply]
Cool, thanks for the quick response!

**Reply by pfebrer** [reply]
I'm doing some stupid test where the `node_attrs` are two times the one hot encoding. I.e. for a dataset of two species, a batch with one atom is `[[0, 1, 0, 1]]` instead of `[[0, 1]]`.

It works until the forward call to  `EquivariantProductBasisBlock`. There `num_elements` is 2, but now `node_attrs` irreps is `4x0e`, and it fails with the message: 

```
RuntimeError: contracted dimensions need to match, but first has size 4 in dim 1 and second has size 2 in dim 0
```

Is the `num_elements` attribute really meant to be `num_elements` or is it the size of `node_attrs` irreps?

**Reply by pfebrer** [reply]
Or should I pass only the one hot encoding to the `node_attrs` argument of `EquivariantProductBasisBlock.forward` ? :thinking:

**Reply by ilyes319** [reply]
yes exactly the product basis will only accept one hot of elements.

**Reply by pfebrer** [reply]
What about the `node_attrs` argument of the interaction? Should it also be only the one hot of elements?

I see that different interaction classes use `node_attrs` differently.

**Reply by rydeveraumn** [reply]
Hey @pfebrer I know this was about a year ago and I am wondering if you were able to get your additional global variables to work? I also want to add global descriptors to my model.

**Reply by pfebrer** [reply]
Hey, yes I managed to do it but unfortunately I can't find the code that I used :( I'm sorry.

Also the code that I used I guess would be outdated with respect to the current MACE in main.

But I remember in the end it was simple enough, you want to pass the global descriptor as part of `node_attrs` when you call `MACE.forward` . To support that you can copy the code of the MACE module, and:
- In `__init__` make sure that the variable `node_attrs_irreps` contains the real node_attrs_irreps, because in the original code it is hardcoded to the number of elements (I think I decided to just accept `node_attrs_irreps` as an argument in `MACE.__init__`).
- In `forward`, make sure the call to the product ([this line in the code](https://github.com/ACEsuit/mace/blob/d39cc6b5f0f416dbc5eb3462f67544592130076e/mace/modules/models.py#L367-L369)) receives as `node_attrs` only the one hot encoding of the species, i.e. remove your extra global descriptors.

I think that was it, but if there is something else that you need to do at least this will be a good start :) Sorry for not being able to provide the code.

**Reply by rydeveraumn** [reply]
Nice! Thank you so much @pfebrer I think I understand. Just curious I am still learning more about the e3nn library and what it entails. What happens if you pass the extra features through the product?

**Reply by rydeveraumn** [reply]
My updated `MACE.forward` definition looks something like this:

```python
        # Embeddings
        node_features_input = g.ndata["node_attrs"]
        if self.dim_state_feats is not None:
            node_features_input = torch.hstack(
                [g.ndata["node_attrs"], g.ndata["state_attrs"]]
            )

        node_feats = self.node_embedding(node_features_input)
        # node_feats = self.node_embedding(g.ndata["node_attrs"])

        # Need to understand how vectors are computed and to do this need to
        # understand shifts
        # Need pbc_offshift here
        edge_attrs = self.spherical_harmonics(vectors)

        # bond_dist here
        edge_feats = self.radial_embedding(
            lengths, g.ndata["node_attrs"], edge_index, self.atomic_numbers
        )

        # For now we will not have pair repulsion
        if hasattr(self, "pair_repulsion"):
            pair_node_energy = self.pair_repulsion_fn(
                lengths, g.ndata["node_attrs"], edge_index, self.atomic_numbers
            )
            if is_lammps:
                pair_node_energy = pair_node_energy[: lammps_natoms[0]]
        else:
            pair_node_energy = torch.zeros_like(node_e0)

        # Interactions
        node_es_list = [pair_node_energy]
        node_feats_list: List[torch.Tensor] = []

        for i, (interaction, product) in enumerate(
            zip(self.interactions, self.products)
        ):
            node_attrs_slice = node_features_input
            product_node_attrs_slice = g.ndata["node_attrs"]

            if is_lammps and i > 0:
                node_attrs_slice = node_attrs_slice[: lammps_natoms[0]]

            node_feats, sc = interaction(
                node_attrs=node_attrs_slice,
                node_feats=node_feats,
                edge_attrs=edge_attrs,
                edge_feats=edge_feats,
                edge_index=edge_index,
                first_layer=(i == 0),
                lammps_class=lammps_class,
                lammps_natoms=lammps_natoms,
            )

            if is_lammps and i == 0:
                node_attrs_slice = node_attrs_slice[: lammps_natoms[0]]

            node_feats = product(
                node_feats=node_feats, sc=sc, node_attrs=product_node_attrs_slice
            )
            node_feats_list.append(node_feats)
```

I added a `dim_state_feats` argument to init based on the dimension of my global features. If it is not None then I pass the full set of features (node attrs + global features) through the node embedding and the interaction, where the node attrs slice has the same dimension as the new set of features. For product, I pass the node features but node_attrs is the original one hot encoded species.

**Reply by pfebrer** [reply]
The product is not an e3nn tensor product if that is what you mean, it is something like the class defined here: https://github.com/ACEsuit/mace/blob/d39cc6b5f0f416dbc5eb3462f67544592130076e/mace/modules/blocks.py#L295

So the thing is that this class I guess just needs to know which species is each atom, and that's why it takes the original `node_attrs`, it just assumes it is a one hot encoding of the species.

**Reply by ilyes319** [reply]
If you look at the most recent develop branch i already made a mechanism to pass arbitrary scale per atom or per graph features to mace.

Have a look here https://github.com/ACEsuit/mace/blob/main/tests%2Ftest_embedding_train.py

**Reply by rydeveraumn** [reply]
Oh great @ilyes319 thank you for sharing!

**Reply by rydeveraumn** [reply]
I will use this implementation I have the `0.3.13` version so I will clone and utilize this functionality in main.
